pkg/helm/helm.go

```diff
@@ -18,165 +18,163 @@ package helm
+// Initialize initializes a chart repository.
+// TODO(f0rmiga): be able to handle multiple repositories. How can we handle charts with the same
+// name across repositories?
+// TODO(f0rmiga): add a readiness probe for this initialization process. A health endpoint would be
+// enough.
+func (c *Client) Initialize(repoURL string) error {
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: initializing")
+	}
 
-	c.home = helmpath.Home(environment.DefaultHelmHome)
-	klog.V(3).Infof("helm client: helm home: %s", c.home)
-	f, err := repo.LoadRepositoriesFile(c.home.RepositoryFile())
+	// TODO(f0rmiga): Allow private repos with authentication. Entry will need to contain the auth
+	// configuration.
+	chartCfg := repo.Entry{
+		Name: "stable",
+		URL:  repoURL,
+	}
+	if chartCfg.URL == "" {
+		chartCfg.URL = stableURL
+	}
+	chartRepo, err := c.repositoryClient.Initialize(&chartCfg, getter.All(c.settings))
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	klog.V(3).Infof("helm client: caching stable repository")
-	cif := c.home.CacheIndex("stable")
-	cr := repo.Entry{
-		Name:  "stable",
-		Cache: cif,
-		URL:   c.repoURL,
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: downloading index file")
 	}
-
-	var settings environment.EnvSettings
-	r, err := repo.NewChartRepository(&cr, getter.All(settings))
+	indexPath, err := c.repositoryClient.DownloadIndex(chartRepo)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	if err := r.DownloadIndexFile(c.home.Cache()); err != nil {
-		return errors.Wrapf(err, "Looks like %q is not a valid chart repository or cannot be reached", cr.URL)
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: loading repository")
+	}
+	indexFile, err := c.repositoryClient.Load(indexPath)
+	if err != nil {
+		return fmt.Errorf("failed to initialize helm client: %v", err)
 	}
 
-	f.Update(&cr)
-	f.WriteFile(c.home.RepositoryFile(), 0644)
-
-	// Load the repositories.yaml
-	c.rf, err = repo.LoadRepositoriesFile(c.home.RepositoryFile())
+	chartRepo.IndexFile = indexFile
+	c.SetChartRepo(chartRepo)
 
-	klog.V(5).Infof("helm client: initialized client")
+	if l := c.log.V(3).Get(); l != nil {
+		l.Log("helm client: successfully initialized")
+	}
 
-	return err
+	return nil
 }
 
-func (c *Client) ListCharts() (map[string]repo.ChartVersions, error) {
-	klog.V(4).Infof("helm client: listing charts")
-
-	charts := map[string]repo.ChartVersions{}
-
-	// TODO: handle non-unique names across repos
-	for _, r := range c.rf.Repositories {
-		n := r.Name
-		f := c.home.CacheIndex(n)
-		index, err := repo.LoadIndexFile(f)
-		if err != nil {
-			return nil, errors.Wrapf(err, "Could not load helm repository index at %s", f)
-		}
+// SetChartRepo allows setting the chart repository being used by the client.
+func (c *Client) SetChartRepo(chartRepo *repo.ChartRepository) {
+	c.chartRepo = chartRepo
```

Comment:
Why isn't `chartRepo` just a public member then? Having it writable seems like a weird break in encapsulation… (So I'm actually advocating for removing this alltogether, but I think that means making it harder to write the tests.)

I guess the actual problem is that `Initialize()` is too big?

Reply:
Fixed with 14f6a35.